### ISSUE TITLE
QWEN 3B I493E

### ISSUE THREAD

**Issue by abdonejam2** [QwenLM/Qwen3-VL#1877]
`
**#Install Dependecies**
!pip install unsloth
!pip install transformers==4.57.1
!pip install --no-deps trl==0.22.2
`

`
**#Loading The Model from Unsloth**
from unsloth import FastVisionModel # FastLanguageModel for LLMs
import torch

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit", # Llama 3.2 vision support
    "unsloth/Llama-3.2-11B-Vision-bnb-4bit",
    "unsloth/Llama-3.2-90B-Vision-Instruct-bnb-4bit", # Can fit in a 80GB card!
    "unsloth/Llama-3.2-90B-Vision-bnb-4bit",

    "unsloth/Pixtral-12B-2409-bnb-4bit",              # Pixtral fits in 16GB!
    "unsloth/Pixtral-12B-Base-2409-bnb-4bit",         # Pixtral base model

    "unsloth/Qwen2-VL-2B-Instruct-bnb-4bit",          # Qwen2 VL support
    "unsloth/Qwen2-VL-7B-Instruct-bnb-4bit",
    "unsloth/Qwen2-VL-72B-Instruct-bnb-4bit",

    "unsloth/llava-v1.6-mistral-7b-hf-bnb-4bit",      # Any Llava variant works!
    "unsloth/llava-1.5-7b-hf-bnb-4bit",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastVisionModel.from_pretrained(
    "unsloth/Qwen3-VL-2B-Instruct-unsloth-bnb-4bit",
    load_in_4bit = True, # Use 4bit to reduce memory use. False for 16bit LoRA.
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for long context
)
`

`
**#Defining Model Parameters**
model = FastVisionModel.get_peft_model(
    model,
    finetune_vision_layers     = True, # False if not finetuning vision layers
    finetune_language_layers   = True, # False if not finetuning language layers
    finetune_attention_modules = True, # False if not finetuning attention layers
    finetune_mlp_modules       = True, # False if not finetuning MLP layers

    r = 16,           # The larger, the higher the accuracy, but might overfit
    lora_alpha = 16,  # Recommended alpha == r at least
    lora_dropout = 0,
    bias = "none",
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
    # target_modules = "all-linear", # Optional now! Can specify a list if needed
)
`

`
**#Preparing and Verifying The Dataset**
import pandas as pd
import os
from PIL import Image as PILImage

# --- 1. Define Paths and Columns ---
CSV_PATH = "/kaggle/input/qwen-dataset/qwen_dataset.csv"
IMAGE_BASE_PATH = "/kaggle/input/comp-data/custom_dataset/train/"

# !! UPDATE THESE COLUMN NAMES BASED ON YOUR CSV FILE !!
FILENAME_COLUMN = 'image' # Column in CSV with image filename (e.g., '0001.jpg')
TARGET_TEXT_COLUMN = 'text'  # Column in CSV with the target text (e.g., 'A cat on a mat.')

# --- 2. Load and Pre-process Metadata ---
try:
    df = pd.read_csv(CSV_PATH)
    print(f"✅ Successfully loaded CSV with {len(df)} entries.")
except FileNotFoundError:
    raise FileNotFoundError(f"Error: CSV file not found at {CSV_PATH}. Check your input path.")

# Check for required columns
if FILENAME_COLUMN not in df.columns or TARGET_TEXT_COLUMN not in df.columns:
    raise ValueError(f"CSV must contain columns '{FILENAME_COLUMN}' and '{TARGET_TEXT_COLUMN}'.")

# --- 3. Custom Formatting Function for QWEN3-VL ---
def create_vl_conversation(row):
    """
    Creates the QWEN3-VL ChatML conversation structure for one row.
    This structure is: [USER (text + image), ASSISTANT (target text)]
    """
    
    # 3a. Construct the full image path
    image_filename = str(row[FILENAME_COLUMN])
    full_path = os.path.join(IMAGE_BASE_PATH, image_filename)
    
    try:
        # 3b. Open the image using PIL (Image object is required)
        image_object = PILImage.open(full_path).convert("RGB")
    except FileNotFoundError:
        print(f"⚠️ Warning: Image not found at {full_path}. Skipping this entry.")
        return None # Return None to filter out bad entries
    
    # 3c. Get the target text
    target_text = str(row[TARGET_TEXT_COLUMN])
    
    # 3d. Format into the ChatML 'conversations' structure
    conversation = [
        {
            "role": "user",
            "content": [
                # The prompt / instruction for the model
                { "type": "text", "text": "" },
                # The actual PIL Image object is passed here
                { "type": "image", "image": image_object } 
            ]
        },
        {
            "role": "assistant",
            "content": [
                { "type": "text", "text": target_text }
            ]
        }
    ]
    return conversation

# --- 4. Apply the Formatting and Convert to List ---
# Apply the function to each row
df['conversations'] = df.apply(create_vl_conversation, axis=1)

# Filter out rows where the image could not be loaded
df_cleaned = df.dropna(subset=['conversations'])

# The final dataset must be a standard Python list of dictionaries/conversations
# Unsloth's SFTTrainer can accept this list directly for the `train_dataset` argument.
final_data_list = df_cleaned['conversations'].tolist()

print(f"\n✅ Cleaned data prepared: {len(final_data_list)} conversation examples.")
print("\n--- Example Conversation Structure ---")
import json
print(json.dumps(final_data_list[0], indent=2, default=str)) # default=str handles the PIL Image object
`
`
**#Generating 1 Output To Verify Performance**

# 1. Enable for inference (as you had it)
FastVisionModel.for_inference(model) # Assuming this function is correctly imported/defined

# --- 2. Data Preparation for Inference ---
# Select the 3rd sample (index 2) from your prepared list
sample = final_data_list[2]

# The instruction/prompt is inside the user turn of the sample
# We need to *re-insert* a specific instruction here because your training instruction was empty: { "type": "text", "text": "" }
instruction = "Answer the following question based on the image. Your response MUST begin with a single word (e.g., 'Yes', 'No', 'Blue', '3'). Immediately follow this word with the pipe symbol '|' and then provide the full, detailed reasoning for your answer."

# Extract the PIL Image object from the conversation structure:
# sample is a list of two dicts (User, Assistant)
# sample[0] is the User dict
# sample[0]['content'] is the list of content elements [Text, Image]
# sample[0]['content'][1] is the Image dict
# sample[0]['content'][1]['image'] is the actual PIL Image object
image_object = sample[0]['content'][1]['image']

# The inference messages list is slightly different from the training format.
# We must use the *extracted* PIL image object directly in the tokenizer call,
# and structure the text part for the chat template.

# Construct the messages list with the new, specific instruction
messages = [
    {"role": "user", "content": [
        # Note: We don't put the image object here for apply_chat_template,
        # we put a placeholder {"type": "image"} and pass the image separately
        {"type": "image"},
        {"type": "text", "text": instruction}
    ]}
]

# --- 3. Tokenization and Generation ---
input_text = tokenizer.apply_chat_template(messages, add_generation_prompt = True)

# Tokenize the input, passing the PIL image object *as the first argument*
inputs = tokenizer(
    image_object, # Pass the PIL image object directly
    input_text,
    add_special_tokens = False,
    return_tensors = "pt",
).to(model.device) # Use model.device instead of "cuda" for compatibility

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer, skip_prompt = True)

# Generate the output
print("\n--- Model Inference Output ---")
_ = model.generate(
    **inputs, 
    streamer = text_streamer, 
    max_new_tokens = 128,
    use_cache = True, 
    # Use standard sampling parameters for better results
    temperature = 0.7, 
    do_sample = True, 
    top_p = 0.9, 
    top_k = 50 
)`

`
**#Defining Training Parameters**
from unsloth.trainer import UnslothVisionDataCollator
from trl import SFTTrainer, SFTConfig

FastVisionModel.for_training(model) # Enable for training!

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    data_collator = UnslothVisionDataCollator(model, tokenizer), # Must use!
    train_dataset = final_data_list,
    args = SFTConfig(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        #max_steps = 30,
        num_train_epochs = 1, # Set this instead of max_steps for full training runs
        learning_rate = 2e-4,
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.001,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none",     # For Weights and Biases

        # You MUST put the below items for vision finetuning:
        remove_unused_columns = False,
        dataset_text_field = "",
        dataset_kwargs = {"skip_prepare_dataset": True},
        max_length = 2048,
    ),
)
`

`
**#Starting Training and Saving Model Weights (Checkpoints) Afterwards to Carry Progress**
trainer_stats = trainer.train()
# After trainer.train() completes successfully:
trainer.save_model("qwen3_vl_lora_adapters")
tokenizer.save_pretrained("qwen3_vl_lora_adapters")
`
`
**#Code for Generating Submission.csv**
import pandas as pd
import os
from PIL import Image as PILImage
import torch 
from transformers import TextStreamer 

from unsloth import FastVisionModel
import torch

# --- Load the Fine-Tuned Model for Inference ---
# Define the path where your trained weights were saved
LORA_ADAPTERS_PATH = "qwen3_vl_lora_adapters"

model, tokenizer = FastVisionModel.from_pretrained(
    model_name = LORA_ADAPTERS_PATH, # Load the adapter path directly
    max_seq_length = 2048,           # Match max_length from training
    dtype = None,
    load_in_4bit = True,             # Keep 4-bit loading for efficiency
)

# Crucial step: Enable model for inference mode
FastVisionModel.for_inference(model)

# --- 1. Define Paths and Constants ---
TEST_CSV_PATH = "/kaggle/input/comp-data/custom_dataset/test_non_labels.csv"
TEST_IMAGE_BASE_PATH = "/kaggle/input/comp-data/custom_dataset/test/"
OUTPUT_CSV_PATH = "submission_one_word_answer.csv" # Changed name for new version

# Assuming 'model' and 'tokenizer' are loaded globally
# device = model.device 

# --- 2. Load Test Data ---
try:
    test_df = pd.read_csv(TEST_CSV_PATH)
    print(f"✅ Successfully loaded test data with {len(test_df)} samples.")
except FileNotFoundError:
    raise FileNotFoundError(f"Error: Test CSV file not found at {TEST_CSV_PATH}")

# Initialize lists to store results
submission_ids = []
submission_answers = []
submission_explanations = []

# --- 3. Inference Loop ---
print("\n🚀 Starting Inference with One-Word Answer Constraint...")
device = model.device 
# Define the strict instruction format the model MUST follow
# This is the key change to enforce separation
STRICT_INSTRUCTION_FORMAT = (
    "Answer the following question based on the image. Your response MUST begin with a single word (e.g., 'Yes', 'No', 'Blue', '3'). "
    "Immediately follow this word with the pipe symbol '|' and then provide the full, detailed reasoning for your answer."
    "\nQuestion: "
)

for index, row in test_df.iterrows():
    sample_id = row['id']
    image_filename = row['file']
    question_text = row['question']
    
    full_image_path = os.path.join(TEST_IMAGE_BASE_PATH, image_filename)
    
    try:
        image_object = PILImage.open(full_image_path).convert("RGB")
    except FileNotFoundError:
        print(f"⚠️ Warning: Image not found for ID {sample_id}. Skipping.")
        continue 
    
    # --- A. Construct Input for Inference (Updated Prompt) ---
    full_prompt = STRICT_INSTRUCTION_FORMAT + question_text
    
    messages = [
        {"role": "user", "content": [
            {"type": "image"}, 
            {"type": "text", "text": full_prompt}
        ]}
    ]

    input_text = tokenizer.apply_chat_template(messages, add_generation_prompt=True)
    
    inputs = tokenizer(
        image_object, 
        input_text,
        add_special_tokens=False,
        return_tensors="pt",
    ).to(device)

    # --- B. Generate Response ---
    with torch.no_grad():
        output_tokens = model.generate(
            **inputs, 
            max_new_tokens=256, 
            temperature=0.1,   
            do_sample=True,    
            top_p=0.9,
            use_cache=True,
            pad_token_id=tokenizer.eos_token_id
        )

    # Decode and clean the generated text
    prompt_len = inputs['input_ids'].shape[1]
    generated_text = tokenizer.decode(output_tokens[0, prompt_len:], skip_special_tokens=True).strip()

    # --- C. Parse Answer and Explanation (Strict Split) ---
    
    # 1. Attempt to split by the required delimiter '|'
    parts = generated_text.split('|', 1) 
    
    if len(parts) == 2:
        # The split was successful
        full_answer_part = parts[0].strip()
        explanation = parts[1].strip()
        
        # 2. Enforce one-word answer by taking only the first word
        # This handles cases where the model forgets to put the pipe symbol immediately
        answer_words = full_answer_part.split()
        answer = answer_words[0] if answer_words else ""
        
        # 3. If there were extra words, prepend them to the explanation for completeness
        if len(answer_words) > 1:
            # Join the remaining words back and add them to the explanation
            remaining_answer = " ".join(answer_words[1:])
            explanation = remaining_answer + " " + explanation
            
    else:
        # Fallback if the model did not use the pipe separator
        print(f"⚠️ Failed to parse ID {sample_id} using '|'. Using crude split.")
        
        # Crude split: take the first word as answer, rest as explanation
        all_words = generated_text.split()
        answer = all_words[0] if all_words else "N/A"
        explanation = " ".join(all_words[1:]) if len(all_words) > 1 else generated_text
        explanation = "Parsing error (Original: " + generated_text + ") - " + explanation

    # --- D. Store Results ---
    submission_ids.append(sample_id)
    submission_answers.append(answer)
    submission_explanations.append(explanation)
    
    print(f"✅ Processed ID {sample_id}. Answer: {answer}. Explanation starts: {explanation[:30]}...")

# --- 4. Create and Save Submission CSV ---
submission_df = pd.DataFrame({
    'id': submission_ids,
    'answer': submission_answers,
    'explanation': submission_explanations
})

submission_df.to_csv(OUTPUT_CSV_PATH, index=False)

print(f"\n🎉 Submission file saved to: {OUTPUT_CSV_PATH}")
print(f"Final CSV has {len(submission_df)} rows and the columns: {submission_df.columns.tolist()}")
`

---------------------------------------------------------------------------------------------------------------
The following is the code used to convert the train_labels.csv dataset into a manageable format for the model to train on. DO NOT RUN.
`
**#Code for converting Database (DO NOT RUN)**
import pandas as pd
import os
from PIL import Image as PILImage

# --- 1. Define Paths and Columns ---
# These are your file paths on Kaggle
CSV_PATH = "/kaggle/input/qwen-dataset/qwen_dataset.csv"
IMAGE_BASE_PATH = "/kaggle/input/comp-data/custom_dataset/train/"

# !! UPDATE THESE COLUMN NAMES BASED ON YOUR CSV FILE !!
# These variables map the column names in your CSV to their meaning.
FILENAME_COLUMN = 'image' # Column in CSV with image filename (e.g., '0001.jpg')
TARGET_TEXT_COLUMN = 'text'  # Column in CSV with the target text (e.g., 'A cat on a mat.')

# --- 2. Load and Pre-process Metadata ---
try:
    df = pd.read_csv(CSV_PATH)
    print(f"✅ Successfully loaded CSV with {len(df)} entries.")
except FileNotFoundError:
    raise FileNotFoundError(f"Error: CSV file not found at {CSV_PATH}. Check your input path.")

# Check for required columns
if FILENAME_COLUMN not in df.columns or TARGET_TEXT_COLUMN not in df.columns:
    raise ValueError(f"CSV must contain columns '{FILENAME_COLUMN}' and '{TARGET_TEXT_COLUMN}'.")

# --- 3. Custom Formatting Function for QWEN3-VL ---
# This function loads the image and puts it into the VLM's conversational format
def create_vl_conversation(row):
    """
    Creates the QWEN3-VL ChatML conversation structure for one row:
    [USER (Instruction + Image), ASSISTANT (Target Text)]
    """
    
    # 3a. Construct the full image path and load the PIL Image object
    image_filename = str(row[FILENAME_COLUMN])
    full_path = os.path.join(IMAGE_BASE_PATH, image_filename)
    
    try:
        # Load the actual image content
        image_object = PILImage.open(full_path).convert("RGB")
    except FileNotFoundError:
        print(f"⚠️ Warning: Image not found at {full_path}. Skipping this entry.")
        return None 
    
    # 3b. Get the target text
    target_text = str(row[TARGET_TEXT_COLUMN])
    
    # 3c. Format into the ChatML 'conversations' structure
    conversation = [
        {
            "role": "user",
            "content": [
                # The instruction text, which was empty in your final training code
                { "type": "text", "text": "" }, 
                # The crucial PIL Image object
                { "type": "image", "image": image_object } 
            ]
        },
        {
            "role": "assistant",
            "content": [
                { "type": "text", "text": target_text }
            ]
        }
    ]
    return conversation

# --- 4. Apply the Formatting and Convert to List ---
# Apply the function to each row
df['conversations'] = df.apply(create_vl_conversation, axis=1)

# Filter out rows where the image could not be loaded and convert to the final list
df_cleaned = df.dropna(subset=['conversations'])
final_data_list = df_cleaned['conversations'].tolist()
`
